### ISSUE TITLE
Ramsey UUID issue

### ISSUE THREAD

**Issue by hictdeveloper** [iwd-nl/snelstart-php#4]
Dear,

Everytime I create a new sales invoice, I get this error:
Fatal error:  Uncaught Ramsey\Uuid\Exception\InvalidUuidStringException: Invalid UUID string:  in /home/{DOMAIN HERE}/public_html/{FOLDER HERE}/vendor/ramsey/uuid/src/Codec/StringCodec.php:146

The sales invoice is still made, but I can't find a solution to this error.

**Comment by hictdeveloper** [comment]
I think it is easy for you to include the stack trace as well:

Stack trace:
#0 /home/{DOMAIN}/public_html/{SUBDOMAIN}/vendor/ramsey/uuid/src/Codec/StringCodec.php(81): Ramsey\Uuid\Codec\StringCodec-&gt;extractComponents('')

#1 /home/{DOMAIN}/public_html/{SUBDOMAIN}/vendor/ramsey/uuid/src/UuidFactory.php(197): Ramsey\Uuid\Codec\StringCodec-&gt;decode('')

#2 /home/{DOMAIN}/public_html/{SUBDOMAIN}/vendor/ramsey/uuid/src/Uuid.php(644): Ramsey\Uuid\UuidFactory-&gt;fromString('')

#3 /home/{DOMAIN}/public_html/{SUBDOMAIN}/vendor/iwd-nl/snelstart-php/src/Mapper/BoekingMapper.php(143): Ramsey\Uuid\Uuid::fromString(NULL)

#4 /home/{DOMAIN}/public_html/{SUBDOMAIN}/vendor/iwd-nl/snelstart-php/src/Mapper/BoekingMapper.php(85): SnelstartPHP\Mapper\BoekingMapper-<mapBoekingResult(Object(SnelstartPHP\Model\Verkoopbo in /home/arutiscom/domains/arutis.com/public_html/tirecenter/vendor/ramsey/uuid/src/Codec/StringCodec.php on line 146

It looks like I am missing a value for creating the Uuid??

**Comment by manuelderuiter** [comment]
Can you give me some more information/details? For example both variables $verkoopboeking and $data from BoekingMapper.php:85?

**Comment by hictdeveloper** [comment]
$verkoopboeking = new \SnelstartPHP\Model\Verkoopboeking();
$verkoopboeking->setKlant($relatie)
    ->setFactuurnummer($new_factnummer)
    ->setFactuurdatum(new \DateTime())
    ->setFactuurbedrag(\Money\Money::EUR($total_incl_btw))
    ->setBoekingsregels($fact_regels)
    ->setBtw([
        (new \SnelstartPHP\Model\Btwregel(\SnelstartPHP\Model\Type\BtwRegelSoort::VERKOPENHOOG(), \Money\Money::EUR($btw)))
    ]);
$boekingConnector->addVerkoopboeking($verkoopboeking);

got no $data variable, could that be it?

**Comment by hictdeveloper** [comment]
So, if that is the case, what should $data be?

**Comment by manuelderuiter** [comment]
As you can see in the stacktrace you are missing some critical information. I believe you passed a ledger that does not exist.

Look at the following line (143): https://github.com/iwd-nl/snelstart-php/blob/43472fdbf8fc735a8f9bd5d22a8efbfa390a23b4/src/Mapper/BoekingMapper.php#L141-L145

It tries to create an instance of 'Grootboek' by its UUID. From the looks of it your Grootboek is not found. If you do see the booking in Snelstart that means we have a issue with the code and the only way we can find out what that issue is, is by dumping $boeking and $data.

https://github.com/iwd-nl/snelstart-php/blob/43472fdbf8fc735a8f9bd5d22a8efbfa390a23b4/src/Mapper/BoekingMapper.php#L114-L121

**Comment by hictdeveloper** [comment]
$verkoopGroot = $grootboekConnector->findByNumber('8002');
Grootboek is alright, the sales invoice is made in Snelstart.

I'll dump the variables and keep you updated.

**Comment by hictdeveloper** [comment]
$boeking dumps:

object(SnelstartPHP\Model\Verkoopboeking)#92 (19) {
  ["klant":"SnelstartPHP\Model\Verkoopboeking":private]=>
  NULL
  ["betalingstermijn":"SnelstartPHP\Model\Verkoopboeking":private]=>
  NULL
  ["eenmaligeIncassoMachtiging":"SnelstartPHP\Model\Verkoopboeking":private]=>
  NULL
  ["doorlopendeIncassoMachtiging":"SnelstartPHP\Model\Verkoopboeking":private]=>
  NULL
  ["bijlagen":protected]=>
  NULL
  ["modifiedOn":protected]=>
  NULL
  ["boekstuk":protected]=>
  NULL
  ["gewijzigdDoorAccountant":protected]=>
  bool(false)
  ["markering":protected]=>
  bool(false)
  ["factuurDatum":protected]=>
  NULL
  ["vervalDatum":protected]=>
  NULL
  ["factuurnummer":protected]=>
  NULL
  ["omschrijving":protected]=>
  NULL
  ["factuurbedrag":protected]=>
  NULL
  ["boekingsregels":protected]=>
  NULL
  ["btw":protected]=>
  NULL
  ["bijlagenUri":protected]=>
  NULL
  ["id":protected]=>
  NULL
  ["uri":protected]=>
  NULL
}

And $data dumps:

array(10) {
  ["verkoopBoeking"]=>
  array(2) {
    ["id"]=>
    string(36) "4cb15301-9689-437e-b7f4-69ba030b7be2"
    ["uri"]=>
    string(54) "/verkoopboekingen/4cb15301-9689-437e-b7f4-69ba030b7be2"
  }
  ["modifiedOn"]=>
  string(22) "2019-04-30T19:13:44.26"
  ["openstaandSaldo"]=>
  float(169.32)
  ["factuurnummer"]=>
  string(5) "10014"
  ["vervalDatum"]=>
  string(23) "2019-05-14T20:13:04.603"
  ["relatie"]=>
  array(2) {
    ["id"]=>
    string(36) "c79dc740-0375-4110-99a2-204a95c9269e"
    ["uri"]=>
    string(46) "/relaties/c79dc740-0375-4110-99a2-204a95c9269e"
  }
  ["factuurDatum"]=>
  string(23) "2019-04-30T20:13:04.603"
  ["factuurBedrag"]=>
  float(169.32)
  ["id"]=>
  string(36) "55793f3b-c2e5-46b7-aff7-05fff7a445bd"
  ["uri"]=>
  string(53) "/verkoopfacturen/55793f3b-c2e5-46b7-aff7-05fff7a445bd"
}

**Comment by hictdeveloper** [comment]
if I dump the $boeking after the $data is put in I get this:

object(SnelstartPHP\Model\Verkoopboeking)#104 (19) {
  ["klant":"SnelstartPHP\Model\Verkoopboeking":private]=>
  NULL
  ["betalingstermijn":"SnelstartPHP\Model\Verkoopboeking":private]=>
  NULL
  ["eenmaligeIncassoMachtiging":"SnelstartPHP\Model\Verkoopboeking":private]=>
  NULL
  ["doorlopendeIncassoMachtiging":"SnelstartPHP\Model\Verkoopboeking":private]=>
  NULL
  ["bijlagen":protected]=>
  NULL
  ["modifiedOn":protected]=>
  object(DateTimeImmutable)#103 (3) {
    ["date"]=>
    string(26) "2019-05-01 13:43:54.043000"
    ["timezone_type"]=>
    int(3)
    ["timezone"]=>
    string(13) "Europe/London"
  }
  ["boekstuk":protected]=>
  NULL
  ["gewijzigdDoorAccountant":protected]=>
  bool(false)
  ["markering":protected]=>
  bool(false)
  ["factuurDatum":protected]=>
  object(DateTimeImmutable)#108 (3) {
    ["date"]=>
    string(26) "2019-05-01 14:43:13.660000"
    ["timezone_type"]=>
    int(3)
    ["timezone"]=>
    string(13) "Europe/London"
  }
  ["vervalDatum":protected]=>
  object(DateTimeImmutable)#105 (3) {
    ["date"]=>
    string(26) "2019-05-01 14:43:13.660000"
    ["timezone_type"]=>
    int(3)
    ["timezone"]=>
    string(13) "Europe/London"
  }
  ["factuurnummer":protected]=>
  string(5) "10035"
  ["omschrijving":protected]=>
  NULL
  ["factuurbedrag":protected]=>
  object(Money\Money)#100 (2) {
    ["amount":"Money\Money":private]=>
    string(5) "15730"
    ["currency":"Money\Money":private]=>
    object(Money\Currency)#107 (1) {
      ["code":"Money\Currency":private]=>
      string(3) "EUR"
    }
  }
  ["boekingsregels":protected]=>
  array(0) {
  }
  ["btw":protected]=>
  array(0) {
  }
  ["bijlagenUri":protected]=>
  NULL
  ["id":protected]=>
  object(Ramsey\Uuid\Uuid)#102 (3) {
    ["codec":protected]=>
    object(Ramsey\Uuid\Codec\StringCodec)#55 (1) {
      ["builder":"Ramsey\Uuid\Codec\StringCodec":private]=>
      object(Ramsey\Uuid\Builder\DefaultUuidBuilder)#56 (1) {
        ["converter":"Ramsey\Uuid\Builder\DefaultUuidBuilder":private]=>
        object(Ramsey\Uuid\Converter\Number\DegradedNumberConverter)#61 (0) {
        }
      }
    }
    ["fields":protected]=>
    array(6) {
      ["time_low"]=>
      string(8) "cbe34b33"
      ["time_mid"]=>
      string(4) "950f"
      ["time_hi_and_version"]=>
      string(4) "4a96"
      ["clock_seq_hi_and_reserved"]=>
      string(2) "98"
      ["clock_seq_low"]=>
      string(2) "ba"
      ["node"]=>
      string(12) "f73e207e1f81"
    }
    ["converter":protected]=>
    object(Ramsey\Uuid\Converter\Number\DegradedNumberConverter)#61 (0) {
    }
  }
  ["uri":protected]=>
  string(53) "/verkoopfacturen/cbe34b33-950f-4a96-98ba-f73e207e1f81"
}

**Comment by hictdeveloper** [comment]
I think I see. The $data['klant'] is empty, but I submitted the addVerkoopboeking with a valid Relation.

Am I correct? How is this possible?

**Comment by hictdeveloper** [comment]
Found problem, when I print_r($boekingsregel); under line 139 I got this:

Array
(
    [omschrijving] => Produktomschrijving
    [grootboek] => 
    [kostenplaats] => 
    [bedrag] => 65
    [btwSoort] => Hoog
)

So, Grootboek is empty in this array, but the invoice is made in Snelstart. I am searching for a solution, but I would be happy if you could think with me.

**Comment by hictdeveloper** [comment]
I can't find anything that leads to the problem that [grootboek] is empty. I hope to hear from you soon. 
Grootboek 8002 is valid and the booking is made in Snelstart, but why is [grootboek] empty then?

Let me know if you need some more information.
Thank you in advance for your help and support.

**Comment by hictdeveloper** [comment]
In addition to my previous comments:

this is what $boeking dumps:

object(SnelstartPHP\Model\Verkoopboeking)#96 (19) {
  ["klant":"SnelstartPHP\Model\Verkoopboeking":private]=>
  NULL
  ["betalingstermijn":"SnelstartPHP\Model\Verkoopboeking":private]=>
  NULL
  ["eenmaligeIncassoMachtiging":"SnelstartPHP\Model\Verkoopboeking":private]=>
  NULL
  ["doorlopendeIncassoMachtiging":"SnelstartPHP\Model\Verkoopboeking":private]=>
  NULL
  ["bijlagen":protected]=>
  NULL
  ["modifiedOn":protected]=>
  object(DateTimeImmutable)#99 (3) {
    ["date"]=>
    string(26) "2019-05-01 14:43:51.317000"
    ["timezone_type"]=>
    int(3)
    ["timezone"]=>
    string(13) "Europe/London"
  }
  ["boekstuk":protected]=>
  NULL
  ["gewijzigdDoorAccountant":protected]=>
  bool(false)
  ["markering":protected]=>
  bool(false)
  ["factuurDatum":protected]=>
  NULL
  ["vervalDatum":protected]=>
  NULL
  ["factuurnummer":protected]=>
  string(5) "10037"
  ["omschrijving":protected]=>
  NULL
  ["factuurbedrag":protected]=>
  NULL
  ["boekingsregels":protected]=>
  NULL
  ["btw":protected]=>
  NULL
  ["bijlagenUri":protected]=>
  NULL
  ["id":protected]=>
  object(Ramsey\Uuid\Uuid)#94 (3) {
    ["codec":protected]=>
    object(Ramsey\Uuid\Codec\StringCodec)#55 (1) {
      ["builder":"Ramsey\Uuid\Codec\StringCodec":private]=>
      object(Ramsey\Uuid\Builder\DefaultUuidBuilder)#56 (1) {
        ["converter":"Ramsey\Uuid\Builder\DefaultUuidBuilder":private]=>
        object(Ramsey\Uuid\Converter\Number\DegradedNumberConverter)#61 (0) {
        }
      }
    }
    ["fields":protected]=>
    array(6) {
      ["time_low"]=>
      string(8) "4b89356a"
      ["time_mid"]=>
      string(4) "9397"
      ["time_hi_and_version"]=>
      string(4) "4936"
      ["clock_seq_hi_and_reserved"]=>
      string(2) "9a"
      ["clock_seq_low"]=>
      string(2) "61"
      ["node"]=>
      string(12) "00b456cc6549"
    }
    ["converter":protected]=>
    object(Ramsey\Uuid\Converter\Number\DegradedNumberConverter)#61 (0) {
    }
  }
  ["uri":protected]=>
  string(53) "/verkoopfacturen/4b89356a-9397-4936-9a61-00b456cc6549"
}

And this is what $data dumps:

array(10) {
  ["verkoopBoeking"]=>
  array(2) {
    ["id"]=>
    string(36) "21a4fc74-56e4-43be-899d-d4d7b84a2cc2"
    ["uri"]=>
    string(54) "/verkoopboekingen/21a4fc74-56e4-43be-899d-d4d7b84a2cc2"
  }
  ["modifiedOn"]=>
  string(23) "2019-05-01T14:43:51.317"
  ["openstaandSaldo"]=>
  float(229.83)
  ["factuurnummer"]=>
  string(5) "10037"
  ["vervalDatum"]=>
  string(23) "2019-05-01T15:43:10.873"
  ["relatie"]=>
  array(2) {
    ["id"]=>
    string(36) "a6b80313-e80b-42fe-8a50-76a88cda6c8b"
    ["uri"]=>
    string(46) "/relaties/a6b80313-e80b-42fe-8a50-76a88cda6c8b"
  }
  ["factuurDatum"]=>
  string(23) "2019-05-01T15:43:10.873"
  ["factuurBedrag"]=>
  float(229.83)
  ["id"]=>
  string(36) "4b89356a-9397-4936-9a61-00b456cc6549"
  ["uri"]=>
  string(53) "/verkoopfacturen/4b89356a-9397-4936-9a61-00b456cc6549"
}

**Comment by manuelderuiter** [comment]
Please do my a favour and create a test case so I can reproduce.

```php
// Start by filling out these variables
$relatie = null;
$new_factnummer = null;
$total_incl_btw = null;
$btw = null;
$fact_regels = [];

$verkoopboeking = new \SnelstartPHP\Model\Verkoopboeking();
$verkoopboeking->setKlant($relatie)
->setFactuurnummer($new_factnummer)
->setFactuurdatum(new \DateTime())
->setFactuurbedrag(\Money\Money::EUR($total_incl_btw))
->setBoekingsregels($fact_regels)
->setBtw([
(new \SnelstartPHP\Model\Btwregel(\SnelstartPHP\Model\Type\BtwRegelSoort::VERKOPENHOOG(), \Money\Money::EUR($btw)))
]);
$boekingConnector->addVerkoopboeking($verkoopboeking);
```

**Comment by hictdeveloper** [comment]
Okay, got it fixed now. Now I'm struggling with adding an PDF attachment :/

Got it like this:

$verkoopboeking = new \SnelstartPHP\Model\Verkoopboeking();
$verkoopboeking->setKlant($relatie)
    ->setFactuurnummer($new_factnummer)
    ->setFactuurdatum(new \DateTime())
    ->setFactuurbedrag(\Money\Money::EUR($total_incl_btw))
    ->setBoekingsregels($fact_regels)
    ->setBtw([
        (new \SnelstartPHP\Model\Btwregel(\SnelstartPHP\Model\Type\BtwRegelSoort::VERKOPENHOOG(), \Money\Money::EUR($btw)))
    ]);
$factuur = $boekingConnector->addVerkoopboeking($verkoopboeking);
$factid = $factuur->getId();

$bijlage = new \SnelstartPHP\Model\VerkoopboekingBijlage();
$bijlage->setVerkoopBoekingId($factid)
    ->setFilename($new_factnummer . '.pdf')
    ->setContent(base64_encode($invoice_pdf));

$boekingConnector->addVerkoopboekingBijlage($factuur, $bijlage);    

but it dumps:
Uncaught GuzzleHttp\Exception\ClientException: Client error: `POST https://b2bapi.snelstart.nl/v2/verkoopboekingen/d6ed097e-cd5e-43c0-a6ad-62020d28a7e4/bijlagen` resulted in a `404 Resource Not Found

I think that it's not the right url for v2, but I'm not sure.

**Comment by Mul-tiMedia** [comment]
> Okay, got it fixed now. Now I'm struggling with adding an PDF attachment :/
> 
> Got it like this:
> 
> $verkoopboeking = new \SnelstartPHP\Model\Verkoopboeking();
> $verkoopboeking->setKlant($relatie)
> ->setFactuurnummer($new_factnummer)
> ->setFactuurdatum(new \DateTime())
> ->setFactuurbedrag(\Money\Money::EUR($total_incl_btw))
> ->setBoekingsregels($fact_regels)
> ->setBtw([
> (new \SnelstartPHP\Model\Btwregel(\SnelstartPHP\Model\Type\BtwRegelSoort::VERKOPENHOOG(), \Money\Money::EUR($btw)))
> ]);
> $factuur = $boekingConnector->addVerkoopboeking($verkoopboeking);
> $factid = $factuur->getId();
> 
> $bijlage = new \SnelstartPHP\Model\VerkoopboekingBijlage();
> $bijlage->setVerkoopBoekingId($factid)
> ->setFilename($new_factnummer . '.pdf')
> ->setContent(base64_encode($invoice_pdf));
> 
> $boekingConnector->addVerkoopboekingBijlage($factuur, $bijlage);
> 
> but it dumps:
> Uncaught GuzzleHttp\Exception\ClientException: Client error: `POST https://b2bapi.snelstart.nl/v2/verkoopboekingen/d6ed097e-cd5e-43c0-a6ad-62020d28a7e4/bijlagen` resulted in a `404 Resource Not Found
> 
> I think that it's not the right url for v2, but I'm not sure.

How exactly did you fix you're problem? I'm having exact the same issues with the Uuid....

**Comment by manuelderuiter** [comment]
Gerben,

There is a difference between the v1 and v2 API. This will work in v1 because it will return an UUID.
```php
$boekingConnector = new \SnelstartPHP\Connector\BoekingConnector($connection);
$verkoopboeking = $boekingConnector->addVerkoopboeking($verkoopboeking);

$bijlage = new \SnelstartPHP\Model\VerkoopboekingBijlage();
$bijlage->setVerkoopBoekingId($verkoopboeking->getId())
    ->setFileName("bestand.pdf")
    ->setContent(base64_encode(file_get_contents("https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf")));
```

I'm working on a revamp of this lib to make it easier to spot the errors...

**Comment by Mul-tiMedia** [comment]
I’m sorry, butt I should have describe my problem more. I’m having issues with the Boekingsregels by setting the grootboek exact the same problem you had in the start post of this issue.

I already found out there is a difference between de v1 and the v2 api :(

**Comment by manuelderuiter** [comment]
Try https://gist.github.com/manuelderuiter/5f9ae9b49c29f5b24eabe389783746cf

**Comment by Mul-tiMedia** [comment]
@manuelderuiter thank you for the examples.
I made some changes to de package, like using the v2 api. And added some extra features.
For uploading documents the v2 works different and has different endpoints.  I managed to upload a file through de v2 api.

Except 1 problem keeps occurring, that is setting the grootboek in the "Boekingsregels", the error keeps popping up: "Invalid UUID string" it looks like grootboek gets emptied, while i added kostenplaats this works flawless. If i set the "grootboek" hardcoded in the Boekingsregels like this:
->setGrootboek(Model\Grootboek::createFromUUID(Uuid::fromString('b41f472d-1c62-4a65-ae63-62b80eec7677')))
https://github.com/Mul-tiMedia/snelstart-php/blob/fa0b67ffa357f8d183e47f7dc2b0fe2306cead29/src/Mapper/BoekingMapper.php#L147

Then it works, except $boekingsregels['grootboek']['id'] keeps popping up empty.... 
Any clues?
